toucan_connectors/net_explorer/net_explorer_connector.py

```diff
@@ -0,0 +1,118 @@
+import json
+import logging
+from io import BytesIO
+from typing import Optional
+
+import pandas as pd
+import requests
+from pydantic import Field, SecretStr
+
+from toucan_connectors.common import ConnectorStatus
+from toucan_connectors.toucan_connector import ToucanConnector, ToucanDataSource
+
+
+class NetExplorerDataSource(ToucanDataSource):
+    file: str
+    sheet: Optional[str] = 0
+
+
+class NetExplorerConnector(ToucanConnector):
+    data_source_model: NetExplorerDataSource
+    instance_url: str = Field(
+        None,
+        Title='Instance URL',
+        placeholder='exemple.netexplorer.pro',
+    )
+    user: str
+    password: SecretStr
+
+    def _retrieve_token(self):
+        login_url = f'https://{self.instance_url}/api/auth'
+
+        data = json.dumps({'user': self.user, 'password': self.password.get_secret_value()})
+
+        headers = {'Content-Type': 'application/json'}
+
+        resp = requests.request('POST', login_url, data=data, headers=headers)
+
+        return resp.json()['token']
+
+    def _retrieve_folders(self, token):
+        folders_url = f'https://{self.instance_url}/api/folders?depth=-1'
+
+        headers = {'Authorization': f'Bearer {token}'}
+
+        resp = requests.request('GET', folders_url, data={}, headers=headers)
+
+        return resp.json()
+
+    def _retrieve_file_id(self, folders, data_source):
+
+        basedir = data_source.file.split('/')[0]
+        path = data_source.file.split('/')[1:]
+
+        id = None
+
+        try:
+            # Search among base directories
+            for folder in folders:
+                if folder['name'] == basedir:
+                    folders = folder['content']
+                    break
+
+            # Serch among paths
+            for elem in path:
+                if elem.endswith(('xlsx', 'xls', 'csv')):
+                    for file in folders['files']:
+                        if file['name'] == elem:
+                            id = file['id']
+                            break
+                else:
+                    for folder in folders['folders']:
+                        if folder['name'] == elem:
+                            folders = folder['content']
+
+            assert id
```

Comment:
Wanted to suggest this, but you can go without it !
```suggestion
        def _search(folders, compare_to, for_id=False):
            for element in folders:
                if element['name'] == compare_to:
                    return element['id'] if for_id else element['content']
        try:
            # Search among base directories
            folders = _search(folders, basedir)
            # Serch among paths
            for elem in path:
                if elem.endswith(('xlsx', 'xls', 'csv')):
                    _id = _search(folders['files'], elem, True)
                else:
                    folders = _search(folders['folders'], elem)
            assert _id
```

Reply:
In case the file doesn't exist :
- My code returned an empty list,
- Your code returns a NoneType object.

**To pass the tests, I added `assert folders` on line 67.**